akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala

```diff
@@ -33,31 +35,74 @@ trait FileUploadDirectives {
    * @group fileupload
    */
   def uploadedFile(fieldName: String): Directive1[(FileInfo, File)] =
+    storeUploadedFile(fieldName, _ ⇒ File.createTempFile("akka-http-upload", ".tmp"))
+
+  /**
+   * Streams the bytes of the file submitted using multipart with the given file name into a designated file on disk.
+   * If there is an error writing to disk the request will be failed with the thrown exception, if there is no such
+   * field the request will be rejected, if there are multiple file parts with the same name, the first one will be
+   * used and the subsequent ones ignored.
+   *
+   * @group fileupload
+   */
+  @ApiMayChange
+  def storeUploadedFile(fieldName: String, destFn: FileInfo ⇒ File): Directive1[(FileInfo, File)] =
     extractRequestContext.flatMap { ctx ⇒
       import ctx.executionContext
       import ctx.materializer
 
       fileUpload(fieldName).flatMap {
         case (fileInfo, bytes) ⇒
 
-          val destination = File.createTempFile("akka-http-upload", ".tmp")
-          val uploadedF: Future[(FileInfo, File)] = bytes.runWith(FileIO.toPath(destination.toPath))
-            .map(_ ⇒ (fileInfo, destination))
+          val dest = destFn(fileInfo)
+          val uploadedF: Future[(FileInfo, File)] = bytes.runWith(FileIO.toPath(dest.toPath))
+            .map(_ ⇒ (fileInfo, dest))
 
           onComplete[(FileInfo, File)](uploadedF).flatMap {
 
             case Success(uploaded) ⇒
               provide(uploaded)
 
             case Failure(ex) ⇒
-              destination.delete()
+              dest.delete()
               failWith(ex)
 
           }
       }
     }
 
   /**
+   * Streams the bytes of the file submitted using multipart with the given field name into designated files on disk.
+   * If there is an error writing to disk the request will be failed with the thrown exception, if there is no such
+   * field the request will be rejected. Stored files are cleaned up on exit but not on failure.
+   *
+   * @group fileupload
+   */
+  @ApiMayChange
+  def storeUploadedFiles(fieldName: String, destFn: FileInfo ⇒ File): Directive1[immutable.Seq[(FileInfo, File)]] =
+    entity(as[Multipart.FormData]).flatMap { formData ⇒
+      extractRequestContext.flatMap { ctx ⇒
+        implicit val mat = ctx.materializer
+        implicit val ec = ctx.executionContext
+
+        val uploaded: Source[(FileInfo, File), Any] = formData.parts
+          .filter(part ⇒ part.filename.isDefined && part.name == fieldName)
```

Comment:
You need to make sure to discard entities for parts that are ignored. Otherwise, big parts that are not files will block the whole stream. Would be nice to add a testcase for that one as well.

Maybe use

```scala
.mapConcat { part =>
  case x if filter(x) => x :: Nil
  case x => 
     x.entity.discardBytes()
     Nil
}
```

Reply:
Ah yep, I see what you mean 👍